File: maximum-product-of-the-length-of-two-palindromic-subsequences/maximum-product-of-the-length-of-two-palindromic-subsequences.java
class Solution {
    int max = 0;
    public int maxProduct(String s) {
        backtrack(s,0,"","");
        return max;
    }
    
    private void backtrack(String s, int i,String s1, String s2){
        if(i == s.length()){
            if(isValid(s1) && isValid(s2)){
                max = Math.max(max, s1.length()*s2.length());
            }
            return;
        }
        backtrack(s,i+1,s1,s2);
        backtrack(s,i+1,s1+s.charAt(i),s2);
        backtrack(s,i+1,s1,s2+s.charAt(i));
    }
    private boolean isValid(String s){
        if(s == null) return false;
        for(int i = 0,j=s.length() - 1;i<j;i++,j--){
            if(s.charAt(i) != s.charAt(j)) return false;
        }
        return true;
    }
}

File: maximum-product-of-the-length-of-two-palindromic-subsequences/maximum-product-of-the-length-of-two-palindromic-subsequences.cpp
class Solution {
    long long ans = 0;
    
    bool isPal(string &s)
    {
        // if(s.size() == 0) return false;
        int start=0, end = s.length()-1;
        while(start < end)
        {
            if(s[start] != s[end]) return false;
            start++; end--;
        }
        return true;
    }
    
    void recur(string &s,int idx,string &s1,string &s2)
    {
        if(idx == s.size())
        {
            if(isPal(s1) and isPal(s2))
            {
                long long val =(int) s1.length() * s2.length();
                ans = max(ans , val);
            }
            return;
        }
        
        s1.push_back(s[idx]);
        recur(s,idx+1,s1, s2);
        s1.pop_back();
        s2.push_back(s[idx]);
        recur(s,idx+1,s1,s2);
        s2.pop_back();
        recur(s,idx+1,s1,s2);
    }
public:
    int maxProduct(string s) {
        string s1="", s2 = "";
        recur(s,0,s1,s2);
        return ans;
    }
};